### PR TITLE
Polygon support

### PR DESCRIPTION
**Pull request by waynenilsen**
Changes to support Polygon/MATIC

Open questions

We need to really figure out what we want to do about finality on Polygon. Based on our conversation with them it sounds like we just take the same approach as we do with Ethereum but that needs to be made into a firm commitment. We are on the hook for sharing our starport specification with them.

Governance of the starport and cash token remain mysterious on Polygon. Eventually of course we want to move this onto  Gateway but the path there is currently somewhat unclear. Perhaps via notices. Needs to be sketched out in detail.

### CODE REVIEW

File: integration/util/scenario/chain_spec.js [hunk 1]
```diff
@@ -34,8 +34,10 @@ async function baseChainSpec(validatorsInfoHash, tokensInfoHash, ctx) {
       eth_address: v.eth_account
   }));
 
-  let assets = tokens.all().map((token) => ({
-    asset: `Eth:${token.ethAddress()}`,
+  let assets = tokens.all()
+      .filter(token => token.symbol.toUpperCase() !== 'CASH')
```
Comment (waynenilsen) [inline integration/util/scenario/chain_spec.js:68]: dev note - i am 99% sure we want this filter but before we had CASH in `SupportedAssets` storage

resulted in having to do some downstream fixes

File: node/src/command.rs [hunk 2]
```diff
@@ -136,6 +136,7 @@ pub fn run() -> sc_cli::Result<()> {
             runtime_interfaces::initialize_validator_config(
                 cli.gateway.eth_key_id.clone(),
                 cli.gateway.eth_rpc_url.clone(),
+                cli.gateway.matic_rpc_url.clone(), // XXX todo: do we even actually need new keys for matic/polygon? I think eth_key_id should be fine / enough.
```
Comment (toninorair) [inline node/src/command.rs]: we probably need it, if infura supports matic as promised, it will be a separate URL for sure.

File: integration/util/scenario/chains.js [hunk 3]
```diff
@@ -46,17 +46,35 @@ class Chain {
     await this.blockInfo.update();
   }
 
+  nameAsPascalCase() {
```
Comment (antonina-cherednichenko) [inline integration/util/scenario/chains.js:49]: maybe just uppercase for all chain names? definitely it looks exotic right now

**Review by toninorair — APPROVED** [review]
Great work, I was a little confused why most of the names involve Matic, but there are still some with Polygon, other than that looks good.

File: ethereum-client/src/events.rs [hunk 4]
```diff
@@ -312,6 +312,8 @@ pub fn decode_event(topics: Vec<String>, data: String) -> Result<EthereumEvent,
         .iter()
         .map(|topic| decode_topic(topic).ok_or(EventError::InvalidTopic))
         .collect::<Result<Vec<ethabi::Hash>, _>>()?;
+
+    // xxx todo:wn note - this is a strange match - need to check the syntax on this and see what it really means.....
```
Comment (jflatow) [inline ethereum-client/src/events.rs]: You mean using the guards?

File: pallets/cash/src/chains.rs [hunk 5]
```diff
@@ -25,13 +25,15 @@ pub enum ChainId {
...
     Dot,
+    Matic, // xxx todo: due to rebrand ticker and name don't match for MATIC/Polygon.. so review this enum name is this what we want to call this?
 }
 
 impl ChainId {
     pub fn to_account(self, addr: &str) -> Result<ChainAccount, Reason> {
         match self {
             ChainId::Gate => Ok(ChainAccount::Gate(Gateway::str_to_address(addr)?)),
             ChainId::Eth => Ok(ChainAccount::Eth(Ethereum::str_to_address(addr)?)),
+            ChainId::Matic => Ok(ChainAccount::Matic(Polygon::str_to_address(addr)?)),
```
Comment (coburncoburn) [inline pallets/cash/src/chains.rs:36]: these should be append only, if Dot has ever been written to storage, it will be wrong next time it is read after this patch. We'd probably get away with it this time, but it's a huge pain if this get's messed up

File: pallets/cash/src/chains.rs [hunk 6]
```diff
@@ -25,13 +25,15 @@ pub enum ChainId {
     Gate,
     Eth,
     Dot,
+    Matic, // xxx todo: due to rebrand ticker and name don't match for MATIC/Polygon.. so review this enum name is this what we want to call this?
```
Comment (jflatow) [inline pallets/cash/src/chains.rs]: Yeah I think this makes sense, not sure what else we would call the network and this still seems to be what they are frequently calling it

File: pallets/cash/src/chains.rs [hunk 7]
```diff
@@ -348,72 +394,138 @@ impl FromStr for ChainId {
 #[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug, Types)]
 pub enum ChainBlock {
     Eth(<Ethereum as Chain>::Block),
+    Matic(<Polygon as Chain>::Block),
```
Comment (coburncoburn) [inline pallets/cash/src/chains.rs:398]: kind of funny to have call it Matic and Polygon on the same line 🤔

File: pallets/cash/src/chains.rs [hunk 8]
```diff
@@ -348,72 +394,138 @@ impl FromStr for ChainId {
...
+        let return_value: Box<dyn Iterator<Item = ChainBlockEvent>> = match self {
+            ChainBlock::Eth(block) => {
+                Box::new(Self::map_events_eth_like(block, ChainBlockEvent::Eth))
+            }
+            ChainBlock::Matic(block) => {
+                Box::new(Self::map_events_eth_like(block, ChainBlockEvent::Matic))
+            }
+        };
+
+        return_value
```
Comment (jflatow) [inline pallets/cash/src/chains.rs]: Yeah its an extra allocate but as a `Vec<ChainBlockEvent>` should be less line noise too

File: pallets/cash/src/chains.rs [hunk 9]
```diff
@@ -348,72 +394,138 @@ impl FromStr for ChainId {
...
+
+    fn blocks_numbers_eth_like(blocks: &Vec<EthereumBlock>) -> impl Iterator<Item = u64> + '_ {
+        blocks.iter().map(|b| b.number)
     }
 
     pub fn block_numbers(&self) -> impl Iterator<Item = u64> + '_ {
-        match self {
-            ChainBlocks::Eth(blocks) => blocks.iter().map(|b| b.number),
-        }
+        let return_value: Box<dyn Iterator<Item = u64>> = match self {
```
Comment (jflatow) [inline pallets/cash/src/chains.rs]: Hm actually surprised it even complains about this one since they are always `u64` items

File: pallets/cash/src/core.rs [hunk 10]
```diff
@@ -319,6 +330,57 @@ pub fn apply_chain_event_internal<T: Config>(event: &ChainBlockEvent) -> Result<
                 result.to_vec(),
             ),
         },
+        ChainBlockEvent::Matic(_block_num, eth_event) => match eth_event {
+            ethereum_client::EthereumEvent::Lock {
```
Comment (coburncoburn) [inline pallets/cash/src/core.rs:334]: maybe these should be refactored to `EthLikeEvent` ?

File: pallets/cash/src/core.rs [hunk 11]
```diff
@@ -354,6 +416,34 @@ pub fn unapply_chain_event_internal<T: Config>(event: &ChainBlockEvent) -> Resul
...
+            _ => Ok(()),
+        },
+        ChainBlockEvent::Matic(_block_num, eth_event) => match eth_event {
+            ethereum_client::EthereumEvent::Lock {
+                asset,
+                sender,
+                chain,
+                recipient,
+                amount,
+            } => internal::lock::undo_lock_internal::<T>(
```
Comment (coburncoburn) [inline pallets/cash/src/core.rs:428]: tbh this boiler plate is a bit unnerving.  copy pasting these same calls for each chain is kind of scary

File: pallets/cash/src/chains.rs (hunk 9 above)
Comment (waynenilsen) [reply pallets/cash/src/chains.rs]: yeah there is a reallly strange error on that one, that is actually the case with a lot of these where the item is actually the same exact type all the way down `EthereumBlock` for example, I couldn't really make sense of the explanation at the time from what i remember. Honestly though this is a very small allocation relative to an entire Vec. Although, they're all quite small allocations overall I guess.. I suppose it is possible that some long lists of blocks could overflow the stack though

File: pallets/cash/src/chains.rs (hunk 7 above)
Comment (waynenilsen) [reply pallets/cash/src/chains.rs:398]: yeah "matic" is the ticker, "matic" is to Polygon as "eth" is to Ethereum

File: ethereum-client/src/events.rs (hunk 4 above)
Comment (waynenilsen) [reply ethereum-client/src/events.rs]: Ok i just did the todo. It turns out that the thing to the left of the `=>` must be a `PATTERN` and here are the docs for this kind of pattern, it is called a "Match Guard".

https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#extra-conditionals-with-match-guards

I _thought_ `t if *t == *LOCK_EVENT_TOPIC` was being evaluated as an expression somehow with that expressions output resulting in the LHS of the match. Importantly `t if *t == *LOCK_EVENT_TOPIC` is NOT in fact an expression and can not be.

**Comment by coburncoburn** [comment]
Looks like a 🧼 implentation of the current architecture, but going through this makes me think we should pre process the blocks and events higher up in the pipeline so more of the runtime code is dealing with `StarportEvent` rather than `EthLikeEvent`, for example. Any thoughts on coming up with an interface for `StarportBlock` etc rather than `EthBlock, FlowBlock`? We should be able to the the branching one time early rather than over and over again.

**Comment by jflatow** [comment]
> Looks like a 🧼 implentation of the current architecture, but going through this makes me think we should pre process the blocks and events higher up in the pipeline so more of the runtime code is dealing with `StarportEvent` rather than `EthLikeEvent`, for example. Any thoughts on coming up with an interface for `StarportBlock` etc rather than `EthBlock, FlowBlock`? We should be able to the the branching one time early rather than over and over again.

I avoided trying to generalize across the chains since most of them we haven't seen yet, and really they are independent. We'll definitely need improvements for 'superblocks' and things at some point, but these kind of *are* the interfaces, just exposed as functions right? Definitely worth discussing

**Comment by waynenilsen** [comment]
Yeah when I was writing the code I was pretty unsatisfied with amount of boilerplate but generally speaking I wanted to implement this feature without major refactoring. 

I think that when we see how Flow fits into the code refactoring will become a bit more obvious and because it is not "eth like" we will see what ends up being actually generic/common and what just seems generic because eth and matic are so darn similar.

File: pallets/cash/src/core.rs (hunk 10 above)
Comment (waynenilsen) [reply pallets/cash/src/core.rs:334]: Yeah the same thing could probably be said for `ethereum_client` and all of the other things in there EthereumBlock etc.. not sure if that makes the situation clearer or not though

**Comment by coburncoburn** [comment]
> > Looks like a 🧼 implentation of the current architecture, but going through this makes me think we should pre process the blocks and events higher up in the pipeline so more of the runtime code is dealing with `StarportEvent` rather than `EthLikeEvent`, for example. Any thoughts on coming up with an interface for `StarportBlock` etc rather than `EthBlock, FlowBlock`? We should be able to the the branching one time early rather than over and over again.
> 
> I avoided trying to generalize across the chains since most of them we haven't seen yet, and really they are independent. We'll definitely need improvements for 'superblocks' and things at some point, but these kind of _are_ the interfaces, just exposed as functions right? Definitely worth discussing

I think the first time we went through this it was more like "ChainThing.do_something" where ChainThing is an trait that EthThing and FlowThing implement and we pass around ChainThing with [dyn](https://doc.rust-lang.org/reference/types/trait-object.html), which would be very close to the enum branches in reality, possible overhead of Boxes and dyn not being *that* cool. 

I'm thinking this time instead we could make gateway very very agnostic of L1's like if the OCW pushes a single type `StarportEvent` into Gateway rather than one of n enum in `ChainEvent`. 

Agreed fits best into superblock discussion

File: pallets/cash/src/core.rs [hunk 12]
```diff
@@ -240,6 +243,14 @@ pub fn recover_validator<T: Config>(
                 }
             }
         }
+        ChainSignature::Matic(eth_sig) => {
```
Comment (toninorair) [inline pallets/cash/src/core.rs:246]: will validator only really produce ChainSignature::Eth? do we really need to add more chains here?

Comment (toninorair) [reply pallets/cash/src/core.rs:246]: @waynenilsen this most likely should go, here is the discussion https://discord.com/channels/402910780124561410/796451735803002900/857390332396568596

File: pallets/runtime-interfaces/src/lib.rs [hunk 13]
```diff
@@ -40,12 +43,14 @@ pub fn initialize_validator_config(
...
 const ETH_KEY_ID_ENV_VAR: &str = "ETH_KEY_ID";
 const ETH_RPC_URL_ENV_VAR: &str = "ETH_RPC_URL";
+const MATIC_RPC_URL_ENV_VAR: &str = "MATIC_RPC_URL";
 const MINER_ENV_VAR: &str = "MINER";
 const OPF_URL_ENV_VAR: &str = "OPF_URL";
 
 const ETH_KEY_ID_DEFAULT: &str = gateway_crypto::ETH_KEY_ID_ENV_VAR_DEV_DEFAULT;
 const MINER_DEFAULT: &str = "Eth:0x0000000000000000000000000000000000000000";
 const ETH_RPC_URL_DEFAULT: &str = "https://ropsten-eth.compound.finance";
+const MATIC_RPC_URL_DEFAULT: &str = "https://ropsten-eth.compound.finance"; // xxx todo:wn fix this before merge
```
Comment (toninorair) [inline pallets/runtime-interfaces/src/lib.rs]: Looks like this needs a fix, especially giving the comment

Comment (waynenilsen) [reply pallets/runtime-interfaces/src/lib.rs]: Yeah, these defaults aren't really terribly useful anyway. I think I'll just set it to `""` and call it a day.

File: pallets/cash/src/core.rs (hunk 12 above)
Comment (waynenilsen) [reply pallets/cash/src/core.rs:246]: when i took another look at this, i completely agree with you, i _think_ I had to put this in there due to the mechanics of the rest of the code (for now) but it was really not what i wanted to do. I'm going to check out what happens if I take it back out but this is a great call out 

I think the other side of the link is here 

`chains.rs:353`

```
pub fn has_validator_signature(&self, chain_id: ChainId, validator: &ValidatorKeys) -> bool {
        match chain_id {
            ChainId::Eth => self.has_signer(ChainAccount::Eth(validator.eth_address)),
            ChainId::Matic => self.has_signer(ChainAccount::Matic(validator.eth_address)),
            _ => false,
        }
    }
```

need to give this some more thought though for sure, it is something that I struggled with in this PR

Comment (waynenilsen) [reply pallets/cash/src/core.rs:246]: also this and various other places 

`chains.rs:361`

```
    pub fn add_validator_signature(
        &mut self,
        signature: &ChainSignature,
        validator: &ValidatorKeys,
    ) -> Result<(), Reason> {
        match (self, signature) {
            (ChainSignatureList::Eth(eth_sig_list), ChainSignature::Eth(eth_sig)) => {
                Ok(eth_sig_list.push((validator.eth_address, eth_sig.clone())))
            }
            (ChainSignatureList::Matic(eth_sig_list), ChainSignature::Matic(eth_sig)) => {
                Ok(eth_sig_list.push((validator.eth_address, eth_sig.clone())))
            }
            _ => Err(Reason::SignatureMismatch),
        }
    }
```

definitely need to think about how we want to deal with this.

Basically, right now, i'm just re-using the same keys for eth and matic since their curve and signing situation is compatible `eth_address` may be something of a misnomer, maybe it should be something more like `eth_like_address` or `evm_compatible_address` etc

File: integration/util/scenario/starport.js [hunk 14]
```diff
@@ -21,8 +21,7 @@ class Starport {
   }
 
   chainAddressStr() {
-    // xxx todo:wn parameterize by chain
-    return `ETH:${this.ethAddress()}`;
+    return `${this.chain.name.toUpperCase()}:${this.ethAddress()}`;
```
Comment (jflatow) [inline integration/util/scenario/starport.js:24]: Still not generalized though, right?

File: integration/util/scenario/starport.js [hunk 15]
```diff
@@ -202,21 +201,21 @@ async function buildStarport(starportInfo, validatorsInfoHash, ctx, chain, cashT
...
     let proxy = await chain.__deploy('TransparentUpgradeableProxy', [
       starportImpl._address,
       proxyAdmin._address,
       "0x"
-    ], { from: chain.root() });
+    ], {from: chain.root()});
     let starport = await chain.__getContractAt('Starport', proxy._address);
     if (validators.length > 0) {
-      await starport.methods.changeAuthorities(validators).send({ from: chain.root(), gas: 4_000_00 });
+      await starport.methods.changeAuthorities(validators).send({from: chain.root(), gas: 4_000_00});
```
Comment (jflatow) [inline integration/util/scenario/starport.js]: Is this gas amount missing a `0`?

File: integration/util/scenario/token.js [hunk 16]
```diff
@@ -160,6 +167,10 @@ class Token {
   async totalChainBorrows() {
     return this.toTokenAmount(await this.ctx.getApi().query.cash.totalBorrowAssets(this.toChainAsset()));
   }
+
+  async getChainBalance(actor) {
+
```
Comment (hayesgm) [inline integration/util/scenario/token.js]: Unused?

File: integration/util/scenario/starport.js (hunk 14 above)
Comment (waynenilsen) [reply integration/util/scenario/starport.js:24]: Yeah starport.js is quite "eth like" specific right now, that refactor will most likely come with flow

File: integration/util/scenario/starport.js (hunk 15 above)
Comment (waynenilsen) [reply integration/util/scenario/starport.js]: hmm it does look that way but i actually only accidentally changed whitespace in here, wondering if I should change the underscores or add the zero, for now i'll change the underscore so it doesn't impact functionality (which is working)

File: ethereum/contracts/Starport.sol [hunk 17]
```diff
@@ -35,9 +35,23 @@ contract Starport {
...
+    constructor(ICash cash_, address admin_, bytes32 chainId_, bytes4 chainIdHeader_) {
         cash = cash_;
         admin = admin_;
+        chainId = chainId_;
+        chainIdHeader = chainIdHeader_;
+    }
+
+    function bytes32ToString(bytes32 _bytes32) public pure returns (string memory) {
+        uint8 i = 0;
+        while(i < 32 && _bytes32[i] != 0) {
```
Comment (hayesgm) [inline ethereum/contracts/Starport.sol]: `while<space>(`

File: integration/util/scenario/actor.js [hunk 18]
```diff
@@ -32,11 +32,13 @@ class Actor {
   }
 
   toChainAccount() {
-    return { Eth: this.ethAddress() };
+    const returnValue = {};
+    returnValue[this.chain().nameAsPascalCase()] = this.ethAddress();
```
Comment (hayesgm) [inline integration/util/scenario/actor.js:36]: this might be `localChainAddress()` instead of `ethAddress()`, possibly?

File: integration/util/scenario/actor.js [hunk 19]
```diff
@@ -138,19 +142,28 @@ class Actor {
     return Number(balance) / 1e6;
   }
 
-  async liquidityForToken(token) {
+  async balanceForToken(token) {
     let assetBalance = await this.ctx.getApi().query.cash.assetBalances(token.toChainAsset(), this.toChainAccount());
+    if(assetBalance == 0) {
```
Comment (hayesgm) [inline integration/util/scenario/actor.js]: `if<space>` and `===` to prevent type coercion.

File: integration/__tests__/lock_scen.js [hunk 20]
```diff
@@ -6,7 +6,8 @@ let lock_scen_info = {
   tokens: [
     { token: 'usdc', balances: { ashley: 1000 } }
   ],
-  validators: ['alice', 'bob']
+  validators: ['alice', 'bob'],
+  native: true
```
Comment (hayesgm) [inline integration/__tests__/lock_scen.js:10]: We don't need native here, but I'm not strictly against it, either. Probably should just use the default though, nah?

File: integration/__tests__/matic_scen.js [hunk 21]
```diff
@@ -1,31 +1,120 @@
...
+            await darlene.lock(10000, maticZrx);
+            balance = await darlene.balanceForToken(maticZrx);
+            let liquidity1 = await darlene.liquidity();
+            expect(balance).toEqual(10000);
+            expect(liquidity1).toBeGreaterThan(0);
+            // now we know everything is in order, make the transfer, let's say half our liquidity worth
+            await darlene.transfer(1000, usdc, chuck);
+            expect(await darlene.chainBalance(usdc)).toEqual(-1000);
+            expect(await chuck.chainBalance(usdc)).toEqual(1000);
+            // xxx todo:wn this test is broken but so is its friend, interest is accruing even though time is not "on"
```
Comment (hayesgm) [inline integration/__tests__/matic_scen.js]: We should fix this, right?

Comment (hayesgm) [reply integration/__tests__/matic_scen.js]: Also, we might want to actually account for interest?

File: integration/__tests__/matic_scen.js [hunk 22]
```diff
@@ -1,31 +1,120 @@
...
             name: 'matic',
-            provider: 'ganache', // [env=PROVIDER]
+            provider: 'ganache',
             ganache: {
                 opts: {},
                 web3_port: null
             },
         },
     },
+    native: true, // for log messages - removes wasm:stripped - no need for wasm no forkless upgrade testing here
```
Comment (hayesgm) [inline integration/__tests__/matic_scen.js:25]: This is more for freeze time here. You should have log messages so long as you run `yarn build` which handles the `wasm stripped` concerns already.

File: integration/util/scenario/actor.js [hunk 23]
```diff
@@ -138,19 +142,28 @@ class Actor {
...
+
+    return token.toTokenAmount(assetBalance.toBigInt());
+  }
+
+  async liquidityForToken(token) {
+    let assetBalance = await this.balanceForToken(token);
     let price = await token.getPrice();
     let liquidityFactor = await token.getLiquidityFactor();
 
     if (assetBalance == 0) {
```
Comment (hayesgm) [inline integration/util/scenario/actor.js]: This one should also be `===`

File: integration/util/scenario/chains.js [hunk 24]
```diff
@@ -47,17 +47,44 @@ class Chain {
...
+
   genesisBlock() {
-    return {
-      Eth: {
-        hash: this.blockInfo.hash,
-        parentHash: this.blockInfo.parentHash,
-        number: this.blockInfo.number,
-        events: []
-      }
+    const block = {};
```
Comment (hayesgm) [inline integration/util/scenario/chains.js]:
```js
return {
  [this.nameAsPascalCase()]: {
    hash: this.blockInfo.hash,
    // ...
  }
}```

File: integration/util/scenario/chains.js [hunk 25]
```diff
@@ -47,17 +47,44 @@ class Chain {
...
+      hash: this.blockInfo.hash,
+      parentHash: this.blockInfo.parentHash,
+      number: this.blockInfo.number,
+      events: []
     };
+    return block;
   }
 
+  genesisBlockForChainSpec() {
+    const newBlock = {};
```
Comment (hayesgm) [inline integration/util/scenario/chains.js]: Can do a similar `{ [xxx]: yyy }` pattern here for nicer syntax.

File: integration/util/scenario/chains.js [hunk 26]
```diff
@@ -233,6 +260,10 @@ class Chains {
       deployment.chain.cashToken = deployment.cashToken;
     });
   }
+
+  genesisBlocksForChainSpec() {
+    return this.all().map(chain => chain.genesisBlockForChainSpec());
```
Comment (hayesgm) [inline integration/util/scenario/chains.js:264]: This returns an array of `[ { Eth: {}, Matic: {} ]`, right?

File: integration/util/scenario/chains.js [hunk 27]
```diff
@@ -272,7 +303,7 @@ async function buildChain(chainInfo, ctx) {
   let version = chainInfo.version ? ctx.versions.mustFind(chainInfo.version) : ctx.versions.current;
 
   let blockInfo = new BlockInfo(web3, chainInfo.block_number, ctx);
-  await blockInfo.update();
+  await blockInfo.update('latest');
```
Comment (hayesgm) [inline integration/util/scenario/chains.js]: Little confused on latest here, doesn't this have a `block_number` that may or may not be latest?

File: integration/util/scenario/starport.js [hunk 28]
```diff
@@ -201,21 +200,21 @@ async function buildStarport(starportInfo, validatorsInfoHash, ctx, chain, cashT
...
     if (validators.length > 0) {
-      await starport.methods.changeAuthorities(validators).send({ from: chain.root(), gas: 4_000_00 });
+      await starport.methods.changeAuthorities(validators).send({from: chain.root(), gas: 400_000});
     }
 
     let starportTopics = Object.fromEntries(starport
-      ._jsonInterface
-      .filter(e => e.type === 'event')
-      .map(e => [e.name, e.signature]));
+        ._jsonInterface
```
Comment (hayesgm) [inline integration/util/scenario/starport.js]: I think we're using 2-spaces here?

File: pallets/cash/src/chains.rs [hunk 29]
```diff
@@ -1004,18 +1265,28 @@ impl Chain for Polkadot {
...
 
 pub fn get_chain_account(chain: String, recipient: [u8; 32]) -> Result<ChainAccount, Reason> {
-    match &chain.to_ascii_uppercase()[..] {
-        "ETH" => {
-            let mut eth_recipient: [u8; 20] = [0; 20];
-            eth_recipient[..].clone_from_slice(&recipient[0..20]);
+    let chain_account_fn = match &chain.to_ascii_uppercase()[..] {
+        "ETH" => ChainAccount::Eth,
+        "MATIC" => ChainAccount::Matic,
+        "MAT" => ChainAccount::Matic,
```
Comment (hayesgm) [inline pallets/cash/src/chains.rs:1286]: Curious about the either here.

File: integration/__tests__/lock_scen.js (hunk 20 above)
Comment (waynenilsen) [reply integration/__tests__/lock_scen.js:10]: this was before we were doing wasm logging was impossible to tell what was going on, also is native faster?

File: pallets/cash/src/chains.rs (hunk 29 above)
Comment (waynenilsen) [reply pallets/cash/src/chains.rs:1286]: this is the whole `bytes4` thing :( i use `MAT:` for the magic header and MATIC for addresses etc. Do you want me to rename the other ones to `MAT` just seemed odd since `eth` is the ticker for Ethereum and `MATIC` is the ticker for Polygon

We can really name these whatever we want but i think it might be nice to use the tickers since they're already defacto market imposed short names on each chain we might bridge

File: integration/util/scenario/actor.js (hunk 18 above)
Comment (waynenilsen) [reply integration/util/scenario/actor.js:36]: Yeah, I think for now having ethAddress is ok but we will ultimately change a lot of this code when flow comes into play. To be honest, I'm a little bit intimidated refactoring something as pervasive as that could be in js without the help of a compiler

File: integration/__tests__/matic_scen.js (hunk 21 above)
Comment (waynenilsen) [reply integration/__tests__/matic_scen.js]: Yeah I can work on this test and its friend

File: integration/util/scenario/chains.js (hunk 24 above)
Comment (waynenilsen) [reply integration/util/scenario/chains.js]: ah thank you!

File: integration/util/scenario/chains.js (hunk 26 above)
Comment (waynenilsen) [reply integration/util/scenario/chains.js:264]: Yes

File: integration/util/scenario/chains.js (hunk 27 above)
Comment (waynenilsen) [reply integration/util/scenario/chains.js]: good find 🦅 👁️ this is a vestigial parameter

File: integration/__tests__/matic_scen.js (hunk 21 above)
Comment (waynenilsen) [reply integration/__tests__/matic_scen.js]: TL;DR - it turns out that the problem is that I was building without the `integration` feature. 

I was doing this because I generally preferred to use debug profile. I mostly have this preference because as I work on rust code I often `cargo build` when i'm ready. If I then `cargo build --release` as in `yarn build` it ends up having to build everything again from scratch. 

Compile times are pretty painful which lead me down this road generally. Cargo check isn't too much better because ultimately you're going to want a build and when you're ready for a build it needs to do the process from scratch anyway. At least rust has incremental builds so that when you don't touch a lot of the code it doesn't have to rebuild much.

File: integration/__tests__/lock_scen.js (hunk 20 above)
Comment (hayesgm) [reply integration/__tests__/lock_scen.js:10]: TBH, it doesn't really matter here. Native is when you're asking for a different genesis native, e.g. running m14's native build. It's irrelevant here, but also, again, if we wanted to change it, we should change the default, not one one-off scen.

File: integration/util/scenario/actor.js (hunk 18 above)
Comment (hayesgm) [reply integration/util/scenario/actor.js:36]: `Cmd+Shift+F` and JavaScript are sadly good friends.

Comment (hayesgm) [reply integration/util/scenario/actor.js:36]: FWIW, we could use TypeScript, but it's a huge burden for tests since it takes a lot more time to write and tests are pretty one-offy. We could technically upgrade if we think it would help us.
